### PR TITLE
Port scala#4737 and scala#4944

### PR DESCRIPTION
**Pull request by nicolasstucki**
Port:
* https://github.com/scala/scala/pull/4737
* https://github.com/scala/scala/pull/4944

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Testing this in https://github.com/lampepfl/dotty/pull/3391

**Comment by nicolasstucki** [comment]
Passed tests in lampepfl/dotty#3391

File: test/junit/scala/tools/nsc/backend/jvm/StringConcatTest.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package scala.tools.nsc
```
Comment (smarter) [inline test/junit/scala/tools/nsc/backend/jvm/StringConcatTest.scala:1]: It'd be nice to copy this test to dotty.

File: src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala [hunk 2]
```diff
@@ -0,0 +1,907 @@
+/* NSC -- new Scala compiler
```
Comment (smarter) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala]: No need to get this class in since we don't use optimizations, don't you think?

**Comment by smarter** [comment]
When porting commits, please put a link to the original PR and put the hash of the original commit in the commit message.

File: test/junit/scala/tools/nsc/backend/jvm/StringConcatTest.scala (hunk 1 above)
Comment (nicolasstucki) [reply test/junit/scala/tools/nsc/backend/jvm/StringConcatTest.scala:1]: Ported in https://github.com/lampepfl/dotty/pull/3391

File: src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala (hunk 2 above)
Comment (nicolasstucki) [reply src/compiler/scala/tools/nsc/backend/jvm/opt/BoxUnbox.scala]: Removed

**Comment by nicolasstucki** [comment]
Passed tests in lampepfl/dotty#3391

**Review by smarter — APPROVED** [review]

**Pull request merged**